src/lua/datetime.lua

```diff
@@ -172,11 +293,96 @@ local function check_ymd_range(y, M, d)
         return
     end
 ::min_err::
-    error(('date %d-%02d-%02d is less than minimum allowed %d-%02d-%02d'):
-        format(y, M, d, MIN_DATE_YEAR, MIN_DATE_MONTH, MIN_DATE_DAY))
+    error(('date %d-%02d-%02d is less than minimum allowed %s'):
+        format(y, M, d, MIN_DATE_TEXT))
 ::max_err::
-    error(('date %d-%02d-%02d is greater than maximum allowed %d-%02d-%02d'):
-        format(y, M, d, MAX_DATE_YEAR, MAX_DATE_MONTH, MAX_DATE_DAY))
+    error(('date %d-%02d-%02d is greater than maximum allowed %s'):
+        format(y, M, d, MAX_DATE_TEXT))
+end
+
+-- check v value against maximum/minimum possible values
+-- if there is nil then simply return default 0
+local function checked_max_value(v, max, txt)
+    if v == nil then
+        return 0
+    end
+    if type(v) ~= 'number' then
+        error(('numeric value expected, but received %s'):
+              format(type(v)), 2)
+    end
+    if v > -max and v < max then
+        return v
+    end
+    error(('value %s of %s is out of allowed range [%s, %s]'):
+            format(v, txt, -max, max), 4)
+end
+
+
+local function bool2int(b)
+    return b and 1 or 0
+end
+
+local function normalize_nsec(secs, nsec)
+    if nsec < 0 or nsec >= NANOS_PER_SEC then
+        secs = secs + math_floor(nsec / NANOS_PER_SEC)
+        nsec = nsec % NANOS_PER_SEC
+    end
+    return secs, nsec
+end
+
+local adjust_xlat = {
+    none = builtin.DT_LIMIT,
+    last = builtin.DT_SNAP,
+    excess = builtin.DT_EXCESS,
+}
+
+local function interval_decouple_args(obj)
+    if is_interval(obj) then
+        return obj.year, obj.month, obj.sec, obj.nsec, obj.adjust
+    end
+    local year = checked_max_value(obj.year, MAX_YEAR_RANGE, 'year')
+    local month = checked_max_value(obj.month, MAX_MONTH_RANGE, 'month')
+    local adjust = adjust_xlat[obj.adjust] or DEF_DT_ADJUST
+
+    local secs = 0
+    secs = secs + (7 * SECS_PER_DAY) *
+                  checked_max_value(obj.week, MAX_WEEK_RANGE, 'week')
+    secs = secs + SECS_PER_DAY * checked_max_value(obj.day, MAX_DAY_RANGE, 'day')
+    secs = secs + (60 * 60) * checked_max_value(obj.hour, MAX_HOUR_RANGE, 'hour')
+    secs = secs + 60 * checked_max_value(obj.min, MAX_MIN_RANGE, 'min')
+    local sec = checked_max_value(obj.sec, MAX_SEC_RANGE, 'sec')
+    check_integer(sec, 'sec')
+    local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
+    check_integer(nsec, 'nsec')
+    check_integer(usec, 'usec')
+    check_integer(msec, 'msec')
+    local count_usec = bool2int(nsec ~= nil) + bool2int(usec ~= nil) +
+                       bool2int(msec ~= nil)
+    if count_usec > 0 then
+        if count_usec > 1 then
```

Comment:
Isn't `count_usec > 0` and `count_usec > 1` the same as just `count_usec > 1`?

Reply:
Hehe, good point! (There used to be some extra action for `count_usec > 0` but since then disappeared)

Reply:
```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -358,11 +358,9 @@ local function interval_decouple_args(obj)
     check_integer(msec, 'msec')
     local count_usec = bool2int(nsec ~= nil) + bool2int(usec ~= nil) +
                        bool2int(msec ~= nil)
-    if count_usec > 0 then
-        if count_usec > 1 then
-            error('only one of nsec, usec or msecs may be defined '..
-                  'simultaneously', 3)
-        end
+    if count_usec > 1 then
+        error('only one of nsec, usec or msecs may be defined '..
+                'simultaneously', 3)
     end
     secs = secs + (sec or 0)
     nsec = (msec or 0) * 1e6 + (usec or 0) * 1e3 + (nsec or 0)
```